api-gateway/caching-with-redis/modules/redis/src/main/java/io/knotx/example/redis/action/RedisCacheAction.java

```diff
@@ -0,0 +1,158 @@
+package io.knotx.example.redis.action;
+
+import io.knotx.example.redis.util.JsonSerializer;
+import io.knotx.fragments.api.Fragment;
+import io.knotx.fragments.handler.api.Action;
+import io.knotx.fragments.handler.api.domain.FragmentContext;
+import io.knotx.fragments.handler.api.domain.FragmentResult;
+import io.knotx.server.api.context.ClientRequest;
+import io.knotx.server.common.placeholders.PlaceholdersResolver;
+import io.knotx.server.common.placeholders.SourceDefinitions;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.core.net.SocketAddress;
+import io.vertx.redis.client.Redis;
+import io.vertx.redis.client.RedisAPI;
+import io.vertx.redis.client.RedisOptions;
+import io.vertx.redis.client.Response;
+import java.util.Optional;
+
+/**
+ * @see RedisCacheActionFactory
+ */
+public class RedisCacheAction implements Action {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RedisCacheAction.class);
+
+  private static final String DEFAULT_HOST = "localhost";
+  private static final int DEFAULT_PORT = 6379;
+  private static final long DEFAULT_TTL = 60;
+
+  private final Redis redisClient;
+  private final RedisAPI redis;
+  private final JsonObject config;
+  private final String payloadKey;
+  private final long ttl;
+  private final Action doAction;
+
+  RedisCacheAction(Vertx vertx, JsonObject config, Action doAction) {
+    this.config = config;
+    this.doAction = doAction;
+    this.payloadKey = getPayloadKey();
+
+    JsonObject cacheConfig = config.getJsonObject("cache");
+    ttl = cacheConfig != null
+        ? cacheConfig.getLong("ttl", DEFAULT_TTL)
+        : DEFAULT_TTL;
+
+    redisClient = Redis.createClient(vertx, getRedisOptions(config));
+    redis = RedisAPI.api(redisClient);
+  }
+
+  @Override
+  public void apply(FragmentContext fragmentContext, Handler<AsyncResult<FragmentResult>> resultHandler) {
+    String cacheKey = getCacheKey(fragmentContext.getClientRequest());
+
+    redisClient.connect(onConnect -> {

```

Comment:
We create the new connection each time the action is invoked. Can we reuse the connection? What about reconnecting on error?

See:
- https://vertx.io/docs/vertx-redis-client/java/